src/main/java/application/FormController.java

```diff
@@ -162,38 +162,32 @@ public EvaluateXPathResponseBean evaluateXpath(@RequestBody EvaluateXPathRequest
     @ApiOperation(value = "Expand the repeat at the given index")
     @RequestMapping(value = Constants.URL_NEW_REPEAT, method = RequestMethod.POST)
     @ResponseBody
-    public RepeatResponseBean newRepeat(@RequestBody RepeatRequestBean newRepeatRequestBean) throws Exception {
+    public FormEntryResponseBean newRepeat(@RequestBody RepeatRequestBean newRepeatRequestBean) throws Exception {
         log.info("New repeat: " + newRepeatRequestBean);
         SerializableFormSession serializableFormSession = sessionRepo.findOne(newRepeatRequestBean.getSessionId());
         FormSession formEntrySession = new FormSession(serializableFormSession);
-
-        JsonActionUtils.descendRepeatToJson(formEntrySession.getFormEntryController(),
+        JSONObject response = JsonActionUtils.descendRepeatToJson(formEntrySession.getFormEntryController(),
                 formEntrySession.getFormEntryModel(),
                 newRepeatRequestBean.getRepeatIndex());
-
         updateSession(formEntrySession, serializableFormSession);
-
-        JSONObject response = JsonActionUtils.getCurrentJson(formEntrySession.getFormEntryController(),
-                formEntrySession.getFormEntryModel());
-        RepeatResponseBean repeatResponseBean = mapper.readValue(response.toString(), RepeatResponseBean.class);
+        formEntrySession.setSequenceId(formEntrySession.getSequenceId() + 1);
+        FormEntryResponseBean repeatResponseBean = mapper.readValue(response.toString(), FormEntryResponseBean.class);
         log.info("New response: " + repeatResponseBean);
         return repeatResponseBean;
     }
 
     @ApiOperation(value = "Delete the repeat at the given index")
     @RequestMapping(value = Constants.URL_DELETE_REPEAT, method = RequestMethod.POST)
     @ResponseBody
-    public RepeatResponseBean deleteRepeat(@RequestBody RepeatRequestBean repeatRequestBean) throws Exception {
-        SerializableFormSession serializableFormSession = sessionRepo.findOne(repeatRequestBean.getSessionId());
+    public FormEntryResponseBean deleteRepeat(@RequestBody RepeatRequestBean deleteRepeatRequestBean) throws Exception {
+        SerializableFormSession serializableFormSession = sessionRepo.findOne(deleteRepeatRequestBean.getSessionId());
         FormSession formEntrySession = new FormSession(serializableFormSession);
-
         JSONObject resp = JsonActionUtils.deleteRepeatToJson(formEntrySession.getFormEntryController(),
                 formEntrySession.getFormEntryModel(),
-                repeatRequestBean.getRepeatIndex(), repeatRequestBean.getFormIndex());
-
+                deleteRepeatRequestBean.getRepeatIndex(), deleteRepeatRequestBean.getFormIndex());
         updateSession(formEntrySession, serializableFormSession);
-
-        return mapper.readValue(resp.toString(), RepeatResponseBean.class);
+        formEntrySession.setSequenceId(formEntrySession.getSequenceId() + 1);
```

Comment:
it'd almost be nice to have this operate at a slightly higher level as some sort of decorator or filter